infra/Pos/Communication/Relay/Logic.hs

```diff
@@ -227,18 +247,29 @@ propagateData enqueue pm = case pm of
     doHandler contents conv = send conv $ DataMsg contents
 
     irdHandler
-        :: Eq key1 => key1 -> contents1
+        :: ( Eq key1 ) => key1 -> contents1
         -> ConversationActions
-             (InvOrData key1 contents1) (ReqMsg key1) m
+             (InvOrData key1 contents1) (ReqOrRes key1) m
         -> m ()
     irdHandler key conts conv = do
         send conv $ Left $ InvMsg key
         let whileNotK = do
-              rm <- recv conv maxBound
-              whenJust rm $ \ReqMsg{..} -> do
-                if rmKey == key
-                   then send conv $ Right $ DataMsg conts
-                   else whileNotK
+              rm <- recvLimited conv
+              whenJust rm $ \msg -> case msg of
+                  Right _ -> throw UnexpectedResponse
+                  Left (ReqMsg{..}) -> do
+                      -- If they send a mismatched key then we don't give them
+                      -- any data.
+                      -- Dangerous. They also have to realize that they
+                      -- shouldn't expect a DataMsg to come next.
+                      if rmKey /= key
+                         then whileNotK
+                         else do
+                             send conv $ Right $ DataMsg conts
+                             rm' <- recvLimited conv
+                             whenJust rm' $ \msg' -> case msg' of
+                                 Left _ -> throw UnexpectedRequest
+                                 Right (ResMsg{..}) -> return ()
```

Comment:
Ok this is just me not understand the existing code, but what role does `irdHandler` play? I thought `invDataListener` did the receiving side of the Inv/Req/Data/Res conversation and `invReqDataFlowDo` did the sending side. So what is `irdHandler` for?